### ISSUE TITLE
Are layers built by a `--squash`ed build useful?

### ISSUE THREAD

**Issue by elibarzilay** [docker/docker#31983]
AFAICT, when I run a `docker build` with a `--squash` flag, what happens is the build result image is generated as usual, and then a squashed layer is created that is not related to where it came from in any way that I can see.  To put this it more concrete terms:

    $ { echo "FROM ubuntu"; echo "RUN touch /foo"; } > Dockerfile

    $ docker build -t foo --squash .
    [...]
    Successfully built 4bf2055740ef

    $ docker images
    REPOSITORY          TAG                 IMAGE ID            CREATED              SIZE
    <none>              <none>              4bf2055740ef        About a minute ago   130 MB
    foo                 latest              af53f215f703        About a minute ago   130 MB
    ubuntu              latest              0ef2e08ed3fa        3 weeks ago          130 MB

The first hint of a problem here is that `af53` in here is not directly related to `4bf2`.  AFAICT, the only connection is indirect: the comment on `af53` says that it's a merge of `sha256:4bf2...`.  But the actual problem that bothers me is that the `af53` hash is not generated in a consistent way, which means that each time I run the *same* build command again, I get a new squashed image:

    $ docker build -t foo --squash .
    [...]
    Successfully built 4bf2055740ef

    $ docker images
    REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
    foo                 latest              81bb7093176a        8 seconds ago       130 MB
    <none>              <none>              4bf2055740ef        9 minutes ago       130 MB
    <none>              <none>              af53f215f703        9 minutes ago       130 MB
    ubuntu              latest              0ef2e08ed3fa        3 weeks ago         130 MB

This shows that `af53` is now a dangling `<none>` thing.  The problem with this is that if I run a `docker build --squash` on, for example, a push to a repo, I'll need to implement some kind of image garbage collection myself since I need to somehow remove `af53` but keep `4bf2`.

**Context, and how this is a worse problem**

The context for this starts at Issue #12198, specifically [this comment](https://github.com/docker/docker/pull/12198#issuecomment-101385063) from @cpuguy83, describing what looks like the way to use `--squash` to achieve a `MARK`/`SQUASH` functionality.  I went ahead and wrote a small shell tool that does just that: splits a single dockerfile into multiple fragments, and runs a build on each fragment.  Each of these builds has a `--tag` with a temporary tag name and the following fragment gets prepended with a `FROM` line with that tag.  Some builds can use a `--squash` flag which makes it possible to get the `MARK`/`SQUASH` thing.

Now, the problem is that if there is a `--squash` build, and other builds follow it.  Each `--squash` run will get a new ID, and that will kill any chance of caching in the following builds.  Worse, it means that making the resulting image public is problematic, since no caching of existing layers will happen.

The only way I see around this is, for each squashed build: run it with `--squash`, then look through the image list for a squashed image with the same exact comment, and if found, move the tag to it and remove the old image.  I really hope that there's a better way to do that since this is not only complex enough to be pretty fragile, but in addition it means that on the build side I'm wasting cpucycles to do a build possibly with the only purpose of getting its comment and dumping it.  I might look at the actual commands in the dockerfile fragment and look for an image with that, but that puts me in the business of parsing dockerfiles and that's much more than just a quick external tool.

So: is there some way to make the squashed image get the same ID consistently that is not as insane as the above?  If not, then it looks to me like squashing builds is still not a sufficient answer to the many requests for `MARK`/`SQUASH` and similar...

**Comment by thaJeztah** [comment]
ping @cpuguy83 @tonistiigi

**Comment by tonistiigi** [comment]
I'm not sure what is the actual problem you are trying to solve or why you need `--squash` at all(I see that the unsquashed and squashed image in your example have the same size), but ID of squashed image changing on every build seems like a bug that shouldn't happen.

**Comment by cpuguy83** [comment]
Yes this looks like a bug that we are not consulting the cache before deciding to squash.

**Comment by elibarzilay** [comment]
I don't know Go so this is me guessing: it seems to me that the problem is that the current time is put in `newImage` ([here](https://github.com/docker/docker/blob/master/daemon/images.go#L306)) and that value gets [marshaled](https://github.com/docker/docker/blob/master/daemon/images.go#L311) and then it looks like this is [hashed](https://github.com/docker/docker/blob/master/daemon/images.go#L316) to make the new ID.  If this is the case, then maybe the fix is to use the time of the source image instead of the time of the squashing?

(But again, this is just a guess.)


(@tonistiigi, these were just minimized toy examples.  The tool that I'm building is basically a wrapper around `docker build` that deals with a meta-annotated dockerfile and the tool splits it according to these annotations into fragments that are each built in turn.  The upshot of this bug is that any build that follows a `--squash`ed build will not use any caches, and in addition multiple builds mean a trail of junk images left behind and need to be collected manually.)

**Comment by benmishkanian** [comment]
I believe I am running into this problem and it's quite inconvenient. My project is forced to use `--squash` as a workaround for the layer depth limit, and I have time-consuming operations occurring after the squashed stage in the Dockerfile. Because the squashed stage is built every time, the layers after that stage are never cached, which adds a lot of time to every build.

Is there a workaround for this problem?
